### PR TITLE
Update topics repo count when deleting repository

### LINKED ISSUE #10038: Counts of tags are not decreased when deleting a repository
- Gitea version (or commit ref): master

## Description

Hello,

I have recently deleted one of my repositories, and browsed through the database to check if the removal was successful. I have found that the count for the tags that repository was associated with have not been decreased. Further, although the tags were only used for this single repository, the tags were still in the database.

By having a look at the code, it seems that the function `removeTopicFromRepo()` from `/models/topic.go` is never called, and hence the `RepoCount` remains unchanged after a repository is deleted. I believe that when deleting a repository in `DeleteRepository()` from `/models/repo.go`, the function should be called for every tag that repository is associated with.

I'd like to work on a pull request to resolve this issue, but I first wanted to confirm that this is indeed a bug. Also, I'm unsure whether to delete a tag if it has a count of `0`, or to leave it in the database.

Lastly, if this was a bug, all current databases must be regarded as inconsistent. Should we reevaluate the counts for the tags?

### PR DESCRIPTION
**Pull request by lafriks**
Fixes #10038

### CODE REVIEW

File: models/migrations/v126.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		if err != nil {
+			return err
+		}
+		if len(results) == 0 {
+			break
+		}
+		last += len(results)
+
+		for _, res := range results {
+			_, err := x.ID(res.ID).Cols("repo_count").
```
Comment (6543) [inline models/migrations/v126.go]: Why not use a session for this migration function?

File: models/migrations/v126.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	_, err := x.Exec(builder.Delete(builder.NotIn("`repo_id`", builder.Select("`id`").From("`repository`"))).From("`repo_topic`"))
+	if err != nil {
+		return err
+	}
+
+	var last int
+	const batchSize = 50
+	for {
```
Comment (lunny) [inline models/migrations/v126.go]: How about directly update set sub query.

File: models/migrations/v126.go (hunk 1 above)
Comment (lafriks) [reply models/migrations/v126.go]: There is not much point in session as data are not related really

File: models/migrations/v126.go (hunk 2 above)
Comment (lafriks) [reply models/migrations/v126.go]: subquery requires ID from updateble table, I don't know if xorm supports such updates

Comment (guillep2k) [reply models/migrations/v126.go]: How about?
```
_, err := x.Exec("UPDATE topic SET repo_count = (SELECT COUNT(*) FROM repo_topic WHERE repo_topic.topic_id = topic.id)");
```

Worked for me in PSQL, SQLite, MySQL and MSSQL.

Comment (guillep2k) [reply models/migrations/v126.go]: AFAIU xorm does not support `INSERT` or `UPDATE` without going through specific records, as expected from an ORM. Only bulk DML operation supported is `DELETE`.

On second thought, it should be interesting to check whether the following is supported, though:

```
_, err := x.Cols("repo_count").
	SetExpr("repo_count", builder.Select("count(*)").From("repo_topic").Where(
		builder.Eq{"repo_topic.topic_id": "topic.ID"},
	)).Update(new(Topic))
```

Comment (lafriks) [reply models/migrations/v126.go]: I think it would treat `topic.ID` as string value

**Comment by lafriks** [comment]
@lunny @guillep2k updated to single update command using xorm builder.

FYI: `.Where(builder.Eq{"'1'": "1"}))` is needed as xorm builder otherwise generates invalid SQL

File: models/migrations/v126.go [hunk 3]
```diff
@@ -10,40 +10,16 @@ import (
...
-			}
-		}
-	}
-	return nil
+	_, err = x.Exec(builder.Update(
+		builder.Eq{
+			"`repo_count`": builder.Select("count(*)").From("`repo_topic`").Where(builder.Eq{
+				"`repo_topic`.`topic_id`": builder.Expr("`topic`.`id`"),
+			}),
+		}).From("`topic`").Where(builder.Eq{"'1'": "1"}))
```
Comment (guillep2k) [inline models/migrations/v126.go:23]: Interesting!

**Review by guillep2k — APPROVED** [review]

File: models/migrations/v126.go (hunk 3 above)
Comment (zeripath) [reply models/migrations/v126.go:23]: Such cunning evilness

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

File: models/migrations/v126.go (hunk 3 above)
Comment (lafriks) [reply models/migrations/v126.go:23]: I just wanted to see if it can be done :laughing:

**Pull request merged**